### ISSUE TITLE
Fix buffer management not handling exceeding the libuv buffer size

### ISSUE THREAD

**Issue by kellabyte** [haywire/haywire#88]
Haywire isn't using libuv and buffers passed to it correctly. In the case a really large request comes in that requires 2 libuv `alloc_cb()` calls we aren't tracking multiple buffers. 

This surfaces 2 problems:
1. Large requests aren't handled properly.
2. During pipelined benchmarks with `64` pipelined requests the buffer can get filled and cause the same problem. 

These will show up as `non-200` errors.

```
../bin/wrk/wrk --script ./pipelined_get.lua --latency -d 10s -t 10 -c 512 http://server:8000/bar/foo2 -- 64
Running 10s test @ http://server:8000/bar/foo2
  10 threads and 512 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.53ms    7.41ms 138.98ms   90.05%
    Req/Sec   391.46k    73.43k  530.99k    54.40%
  Latency Distribution
     50%    3.57ms
     75%    6.54ms
     90%   12.88ms
     99%    0.00us
  39188258 requests in 10.08s, 6.42GB read
  Socket errors: connect 0, read 8, write 0, timeout 0
  Non-2xx or 3xx responses: 290
Requests/sec: 3888292.59
Transfer/sec:    652.57MB
```

Another way to reproduce is:
```
dd if=/dev/urandom of=test.dat bs=1048576 count=1
curl -D @test.dat http://server:8000 
```

**Comment by kellabyte** [comment]
There are 2 options to solve this.

### 1. Abandon zero-copy optimizations
 memcpy into expanding buffers.

### 2. Continue with zero-copy optimizations
Manage multiple buffers for a single request and retain pointer locations. This means a body (and potentially header?) could span multiple buffers. This means we can't expose pointers to user facing API's and we need to expose 2 types of API's.

1. A streaming API to iterate the bytes in a virtual buffer that spans multiple real buffers.
2. A memcpy API in case someone wants to do `foo = request->get_body()` as a whole value. 

This allows us to retain zero-copy when people don't use certain headers or body values instead of memcpy'ing everything while still offering a memcpy when someone wants the full buffer.

### Concerns
The big question: Does all this buffer management negate the performance optimization of zero-copy and would memcpy just end up being equal and simpler?

**Comment by nmdguerreiro** [comment]
I'm in favour of implementing option 3), as that was what I was planning to try first after thinking about it a bit more. 

I'm assuming that option 1) is the same as 3), except for the fact there would be an additional copy to generate null terminated strings and not to use `hw_string`.

Option 2) isn't very appealing IMHO. Even though we'd be saving CPU time in Haywire land, assuming the client wants all the data in one go (i.e. non-streaming) it would have to pay the CPU cost of iterating through the stream or allocating a big enough memory block and copying the data there. The end result would be that we would be doing memcpy in Haywire.

Going back to option 3), if we allocate 64KB by default and keep track of how much we've used and give the rest back to libuv during the alloc callback, we wouldn't need to realloc unless required.

I'll have a go at this and see how far along I get.

**Comment by kellabyte** [comment]
Hmm interesting points. I don't have any experience using `realloc()`. Pointer locations changing could be a problem. Let me know what you find out.

**Comment by nmdguerreiro** [comment]
Fixed by #103
